service/history/hsm/tree.go

```diff
@@ -700,6 +730,48 @@ func (n *Node) root() *Node {
 	return root
 }
 
+// Compact filters the operation log to remove any transitions for deleted nodes while preserving deletion operations.
+// The returned log maintains deletion tombstones, but excludes transitions that occurred on deleted nodes or their
+// descendants.
+func (ol OperationLog) Compact() OperationLog {
+	var deletedPaths [][]Key
+	result := make(OperationLog, 0, len(ol))
+	hasDeleteOperation := make(map[string]bool)
+
+	// First get deletion paths and record existing deletions
+	for _, op := range ol {
+		if del, ok := op.(DeleteOperation); ok {
+			deletedPaths = append(deletedPaths, del.Path())
+			pathKey := fmt.Sprintf("%v", del.Path())
```

Comment:
I would probably just implement this as a tree too instead of formatting but it's not critical.

Reply:
```go
type opNode struct {
	key Key
	children map[Key]opNode
	isDeleted bool
}
```

Then you can traverse the tree to know whether you're in a deleted branch.

Reply:
See this for reference: https://github.com/temporalio/temporal/blob/0b9febbd0632428834589c9bdb597deaaf26648b/tools/testrunner/testrunner.go#L44

Reply:
I _think_ this is what you had in mind. Please take a look.